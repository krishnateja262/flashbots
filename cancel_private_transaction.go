package flashbots

import (
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/rpc"
	"github.com/lmittmann/w3/core"
)

type cancelPrivateTransactionRequest struct {
	TxHash common.Hash `json:"txHash"`
}

// CancelPrivateTransaction stops the private transactions with the given hash
// from being submitted for future blocks by the Flashbots relay.
func CancelPrivateTransaction(hash common.Hash) core.CallerFactory[bool] {
	return &cancelPrivateTransactionFactory{hash: hash}
}

type cancelPrivateTransactionFactory struct {
	// args
	hash common.Hash

	// returns
	returns *bool
}

func (f *cancelPrivateTransactionFactory) Returns(success *bool) core.Caller {
	f.returns = success
	return f
}

func (f *cancelPrivateTransactionFactory) CreateRequest() (rpc.BatchElem, error) {
	return rpc.BatchElem{
		Method: "eth_cancelPrivateTransaction",
		Args: []any{&cancelPrivateTransactionRequest{
			TxHash: f.hash,
		}},
		Result: &f.returns,
	}, nil
}

func (f *cancelPrivateTransactionFactory) HandleResponse(elem rpc.BatchElem) error {
	if err := elem.Error; err != nil {
		return err
	}
	return nil
}
