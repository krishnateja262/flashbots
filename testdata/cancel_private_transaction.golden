> {"jsonrpc":"2.0","id":1,"method":"eth_cancelPrivateTransaction","params":[{"txHash":"0x45df1bc3de765927b053ec029fc9d15d6321945b23cac0614eb0b5e61f3a2f2a"}]}
< {"jsonrpc":"2.0","id":1,"result":true}
